xds/src/main/java/io/grpc/xds/XdsClientImpl.java

```diff
@@ -0,0 +1,721 @@
+          } else if (typeUrl.equals(ADS_TYPE_URL_RDS)) {
+            handleRdsResponse(response);
+          }
+          // TODO(zdapeng): add CDS/EDS response handles.
+        }
+      });
+    }
+
+    @Override
+    public void onError(final Throwable t) {
+      syncContext.execute(new Runnable() {
+        @Override
+        public void run() {
+          handleStreamClosed(
+              Status.fromThrowable(t).augmentDescription("ADS stream [" + this + "] had an error"));
+        }
+      });
+    }
+
+    @Override
+    public void onCompleted() {
+      syncContext.execute(new Runnable() {
+        @Override
+        public void run() {
+          handleStreamClosed(
+              Status.UNAVAILABLE.withDescription("ADS stream [" + this + "] was closed"));
+        }
+      });
+    }
+
+    private void handleStreamClosed(Status error) {
+      checkArgument(!error.isOk(), "unexpected OK status");
+      if (closed) {
+        return;
+      }
+      closed = true;
+      cleanUp();
+      configWatcher.onError(error);
+      if (responseReceived || retryBackoffPolicy == null) {
+        // Reset the backoff sequence if had received a response, or backoff sequence
+        // has never been initialized.
+        retryBackoffPolicy = backoffPolicyProvider.get();
+      }
+      long delayNanos = 0;
+      if (!responseReceived) {
+        delayNanos =
+            Math.max(
+                0,
+                retryBackoffPolicy.nextBackoffNanos() - stopwatch.elapsed(TimeUnit.NANOSECONDS));
+      }
+      logger.log(Level.FINE, "{0} stream closed, retry in {1} ns", new Object[]{this, delayNanos});
+      if (delayNanos == 0) {
+        startRpcStream();
+        if (configWatcher != null) {
+          adsStream.sendLdsRequest(ldsResourceName);
+        }
+        // TODO(chengyuanzhang): send CDS/EDS requests if CDS/EDS watcher presents.
+      } else {
+        rpcRetryTimer =
+            syncContext.schedule(
+                new RpcRetryTask(), delayNanos, TimeUnit.NANOSECONDS, timeService);
+      }
+    }
+
+    private void close(Exception error) {
+      if (closed) {
+        return;
+      }
+      checkState(pendingAckRequest == null, "Severe bug: pending ACK/NACK not sent");
+      closed = true;
+      cleanUp();
+      requestWriter.onError(error);
+    }
+
+    private void cleanUp() {
+      if (adsStream == this) {
+        adsStream = null;
+      }
+    }
+
+    private void sendLdsRequest(String resourceName) {
+      checkState(requestWriter != null, "ADS stream has not been started");
+      DiscoveryRequest request =
+          DiscoveryRequest
+              .newBuilder()
+              .setVersionInfo(ldsVersion)
+              .setNode(node)
+              .addResourceNames(resourceName)
+              .setTypeUrl(ADS_TYPE_URL_LDS)
+              .setResponseNonce(ldsRespNonce)
+              .build();
+      requestWriter.onNext(request);
+    }
+
+    private void sendRdsRequest(String resourceName) {
```

Comment:
suggestion: can you combine `sendLdsRequest` and `sendRdsRequest` into a single `sendXdsRequest` and pass versionInfo, typeUrl and responseNonce as parameters? It will be good to eliminate code duplication as much as possible. In fact your `prepareAckRequest` is almost the function I am talking about. So all 3 can call a single method and pass appropriate values to it.

Reply:
Sure. Combined `sendLdsRequest` and `sendRdsRequest` into `sendXdsRequest`. Also eliminated the logic for first preparing an ACK request and then send it out. Instead, we do send ACK or NACK without preparing, so we do not need to maintain a state for that.